### DISCUSSION TITLE
declare/create type alias

### DISCUSSION THREAD

**Discussion by higelino** [PyO3/pyo3#2683] [Questions]
Is there a way to declare a type alias ? Something equivalent to what follows:

MyUnion = Union[MyPyRustType1, MyPyRustType2]

or 

MyTuple = Tuple[MyPyRustType1, MyPyRustType2]

**Comment by davidhewitt** [comment]
Can you clarify your use case?

**Reply by daemontus** [reply]
I know this is an old question, but I would actually like to know the answer, so I think I can try to provide more details :)

1. In "normal Python", one can write that `TypeX: TypeAlias = TypeY`. Now, someone can import this type alias (`from my_library import TypeX`) and use it in their type annotations (`def someones_function(x: TypeX)`).
2. In PyO3, I would *typically* handle all type annotations through a `.pyi` stub file. This works fine for almost anything. However, if I declare a type alias in my stub file, it can be used in the type signatures, but it is not actually a member of the module and cannot be imported/used by anyone using my library.

So the question is basically... how should I properly create a type alias that is a member of a PyO3 generated module and can be imported by people using my library?

**Comment by daemontus** [comment]
Ok, I think I figured it out:

1. Create a mixed Rust/Python project according to the [maturin guide](https://www.maturin.rs/project_layout).
2. Put your type stubs **and** the type aliases into `__init__.pyi`.
3. Put `from .my_project import *` and `__doc__ = my_project.__doc__` into the `__init__.py`, as suggested in the maturin guide. But do not add the `#if hasattr(my_project, "__all__") ...` part, otherwise the type aliases will still work, but will not show up in your documentation.
5. Also put you type aliases into `__init__.py`, including their docstrings.

With this setup, I now have:
 - Imports correctly resolving in the Python interpreter and in VSCode through `pylance`.
 - `mypy` type checks work, including imported type aliases.
 - `pdoc` shows all Rust classes/methods *and* Python type aliases, including correct docstrings (i.e. Rust docs for Rust code, and Python docs for Python aliases).
